### ISSUE TITLE
Disjoint capture in async blocks requires mut for Copy field modifications

### ISSUE THREAD

**Issue by DangerousVegetable** [rust-lang/rust#146547]
I tried this code:

```rust
use std::future::Future;
use std::pin::Pin;
use std::thread::sleep;
use std::time::Duration;

async fn run_multiple_times<Fut>(
    mut f: impl FnMut() -> Pin<Box<dyn Future<Output = Fut> + Send>>,
) {
    for _ in 0..5 {
        f().await;
        sleep(Duration::from_secs(1));
    }
}

struct Test {
    pub num: usize,
    pub name: String,
}

impl Test {
    fn inc(&mut self) {
        self.num += 1
    }
}

#[tokio::main]
async fn main() {
    let mut state = Test {num: 0, name: "test".to_string()};
    run_multiple_times(move || {
        Box::pin(async move {
            println!("State: {}", state.num);

            state.num = state.num + 1; // This line doesn't capture state, yet requires it to be mut ???

            // These two lines, as expected, do not compile:
            // state.inc() // error: cannot move out of `state`
            // state.name = "test".to_string(); // error: String doesn't implement Copy trait
        })
    }).await;
    println!("State: {}", state.num);
}
```

I expected to see a compilation error due to `state` being moved twice. Instead, this code compiled without any warnings and printed "State: 0" 6 times. This issue also exists on the nightly version.

The way I understand this is that the compiler decides to move `state.num` instead of the whole `state` (which is already confusing) and, since `usize` implements `Copy` trait, it simply copies its value. The problem is that the compiler also requires `state` to be declared as `mut` which is definitely not how it should be.

Am I missing something or is it an actual bug?

**Comment by theemathas** [comment]
My guess is that it's done this way to be consistent with closures. See https://doc.rust-lang.org/edition-guide/rust-2021/disjoint-capture-in-closures.html

Your code has the compile error you expected in edition 2018. The edition guide should probably mention though that it also affects async blocks...

**Comment by DangerousVegetable** [comment]
Thanks for clarification. However, it's still confusing to me why the compiler requires `state` to be `mut`.
